File: TestPublicKeyCryptosystem.java
import java.io.File;
import java.io.IOException;
import java.math.BigInteger;

public class TestPublicKeyCryptosystem {
    public static void main(String[] args) {
        // ElGamal
        System.out.println("ElGamal");
        ElGamal elGamal = new ElGamal();
        BigInteger e_plainText = BigInteger.valueOf(331);
        System.out.println("Plain Text: " + e_plainText);
        BigInteger randomKey = BigInteger.valueOf(197);

        try {
            File output_encrypt = new File("ElGamal_encryptTimes.txt");
            Pair<BigInteger> e_encryptedText = elGamal.timedEncrypt(e_plainText, randomKey, output_encrypt, ',');
            System.out.println("Encrypted Text: " + e_encryptedText);

            File output_decrypt = new File("ElGamal_decryptTimes.txt");
            BigInteger e_decryptedText = elGamal.timedDecrypt(e_encryptedText, output_decrypt, ',');
            System.out.println("Decrypted Text: " + e_decryptedText);
        } catch (IOException e) {
            e.printStackTrace();
        }

        System.out.println();

        // RSA
        System.out.println("RSA");
        RSA rsa = new RSA();
        System.out.println("Public Key: " + rsa.getPublicKey());
        BigInteger r_plainText = BigInteger.valueOf(1070777);
        System.out.println("Plaintext: " + r_plainText);

        try {
            File output_encrypt = new File("RSA_encryptTimes.txt");
            BigInteger r_encryptedText = rsa.timedEncrypt(r_plainText, output_encrypt, ',');
            System.out.println("Encrypted Text: " + r_encryptedText);

            File output_decrypt = new File("RSA_decryptTimes.txt");
            BigInteger r_decryptedText = rsa.timedDecrypt(r_encryptedText, output_decrypt, ',');
            System.out.println("Decrypted Text: " + r_decryptedText);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: C++/Tests/TestPublicKeyCryptosystem.cpp
#include <iostream>
#include <fstream>

#include "ElGamal.h"
#include "RSA.h"
#include "Pair.h"
#include "grabbingRandomPrime.h"

int main(int argc, const char * argv[]) {
    srand(static_cast<int>(time(NULL)));
    
    // ElGamal
    std::cout << "ElGamal" << std::endl;
    ElGamal elGamal = ElGamal();
    std::cout << "Moulus: " << elGamal.getModulus() << std::endl;
    std:: cout << "Primitive Root: " << elGamal.getPrimElement() << std::endl;
    std:: cout << "Public Key: " << elGamal.getPublicKey() << std::endl;
    long long e_plainText = 331;
    std::cout << "Plain Text: " << e_plainText << std::endl;
    long long randomKey = 197;
    
    std::ofstream output;
    
    output.open("ElGamal_encryptionTimes.txt");
    Pair<long long> e_encryptedText = elGamal.timedEncrypt(e_plainText, randomKey, output, ',');
    std::cout << "Encrypted Text: " << e_encryptedText.toString() << std::endl;
    output.close();
    
    output.open("ElGamal_decryptionTimes.txt");
    long long e_decryptedText = elGamal.timedDecrypt(e_encryptedText, output, ',');
    std::cout << "Decrypted Text: " << e_decryptedText << std::endl;
    output.close();
    
    std::cout << std::endl;
    
    // RSA
    std::cout << "RSA" << std::endl;
    RSA rsa = RSA(1223, 1987);
    std::cout << "Modulus: " << rsa.getModulus() << std::endl;
    std::cout << "Public Key: " << rsa.getPublicKey() << std::endl;
    long long r_plainText = 1070777;
    std::cout << "Plaintext: " << r_plainText << std::endl;
    
    output.open("RSA_encryptTimes.txt");
    long long r_encryptedText = rsa.timedEncrypt(r_plainText, output, ',');
    std::cout << "Encrypted Text: " << r_encryptedText << std::endl;
    output.close();

    output.open("RSA_decryptTimes.txt");
    long long r_decryptedText = rsa.timedDecrypt(r_encryptedText, output, ',');
    std::cout << "Decrypted Text: " << r_decryptedText << std::endl;
    output.close();

    return 0;
}
